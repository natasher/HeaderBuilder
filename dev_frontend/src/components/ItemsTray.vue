<template>
  <section id = "mfn-hb-items">

    <div class = "header">

      <span class = "title">
        Drag item and drop into workspace
      </span>

    </div>

    <draggable
      class    = "hb-items clearfix"
      :options = "{
        group: {
          name: 'items',
          sort: false,
          pull: 'clone',
          put : false,
        },
        scroll: true,
      }"
      @start  = "atStart"
      @end    = "atEnd"
      v-model = "baseItemsList">

      <BaseItem
        v-for = "( item, index ) in baseItemsList"
        :key  = "index"
        :item = "item"
        :placeholder = "item.name === draggedItem ? true : false">
      </BaseItem>

    </draggable>

  </section>
</template>

<script>
import { mapActions } from 'vuex'
import FontIcon  from './BaseGeneric/FontIcon.vue'
import BaseItem  from './BaseItem.vue'
import draggable from 'vuedraggable'

export default {

  name: 'ItemsTray',

  data() {
    return {
      draggedItem: '',
    }
  },

  components: {
    FontIcon,
    BaseItem,
    draggable,
  },

  computed: {
    /**
     * Getter method, retrive items boilerplates from store
     *
     * @returns {Object} Items data
     */
    baseItemsList: {
      get() {
        return this.$store.state.items.baseItemsList
      },
      set() {},
    },
  },

  methods: {
    ...mapActions( 'items', [ 'setTemporaryItem' ]),
    /**
     * Set value of private variable `draggedItem` to item name.
     *
     * @param {event} SortableJS
     * @returns {sideEffect} dragged item name
     * @public
     */
    atStart( event ) {
      this.draggedItem = event.item.attributes[ 0 ].nodeValue
      const clonedItem = JSON.parse(JSON.stringify( this.baseItemsList[ event.oldIndex ] ))

      this.setTemporaryItem({ tmpItem: clonedItem })
    },

    /**
     * Clears `draggedItem` variable.
     *
     * @param {event} unused
     * @returns {sideEffect} empty string
     * @public
     */
    atEnd() {
      this.draggedItem = ''
    },
  },
}
</script>
