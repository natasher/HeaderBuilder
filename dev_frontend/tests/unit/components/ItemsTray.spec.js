import { shallowMount, createLocalVue } from '@vue/test-utils'
import Vuex          from 'vuex'
import flushPromises from 'flush-promises'
import merge         from 'lodash.merge'

import ItemsTray     from '../../../src/components/ItemsTray'

const localVue = createLocalVue()
localVue.use( Vuex )

function createStore(overrides) {
  const defaultStoreConfig = {
    modules: {
      items: {
        namespaced: true,
        actions: {
          setTemporaryItem: jest.fn()
        },
        state: {
          baseItemsList: [
            {
              name: 'logo',
              icon: 'heart-line',
            },
            {
              name: 'menu',
              icon: 'list',
            },
            {
              name: 'menuIcon',
              icon: 'menu',
            },
            {
              name: 'extras',
              icon: 'plus-squared',
            },
            {
              name: 'social',
              icon: 'share',
            },
            {
              name: 'text',
              icon: 'doc-text',
            },
            {
              name: 'image',
              icon: 'picture',
            },
            {
              name: 'icon',
              icon: 'feather',
            },
            {
              name: 'button',
              icon: 'db-shape',
            },
          ],
        }
      }
    }
  }

  return new Vuex.Store(
    merge( defaultStoreConfig, overrides )
  )
}

function createWrapper(overrides) {
  const defaultMountingOptions = {
    localVue,
    store: createStore(),
    stubs: [
      'FontIcon',
      'BaseItem',
      'draggable',
    ],
  }

  return shallowMount( ItemsTray, merge( defaultMountingOptions, overrides ) )
}

describe('ItemsTray:', () => {

  describe('snapshots:', () => {

    test('should match when no item is dragged', async () => {
      const wrapper = createWrapper()
      await flushPromises()

      expect( wrapper.element ).toMatchSnapshot()
    })

    test('should match if item is dragged', async () => {
      const wrapper = createWrapper()
      wrapper.setData({ draggedItem: 'extras' })
      await flushPromises()

      expect( wrapper.element ).toMatchSnapshot()
    })

  })

  describe('methods and computed:', () => {

    test('`atStart` should set `data.draggedItem` to given item name if draggable start event is trigger', async () => {
      const event = {
        item: {
          attributes: [{ nodeValue: 'logo' }],
        },
        oldIndex: 0
      }
      const setTemporaryItem = jest.fn()
      const store = createStore({
        modules: {
          items: {
            actions: {
              setTemporaryItem
            }
          }
        }
      })
      const wrapper = createWrapper({ store, })
      await flushPromises()

      wrapper.vm.atStart( event )

      expect( wrapper.vm.draggedItem ).toBe( event.item.attributes[0].nodeValue )
    })

    test('`atStart` should invoke `setTemporaryItem` with cloned item if draggable start event is triggered', async () => {
      const event = {
        item: {
          attributes: [{ nodeValue: 'logo' }],
        },
        oldIndex: 0
      }
      const setTemporaryItem = jest.fn()
      const store = createStore({
        modules: {
          items: {
            actions: {
              setTemporaryItem
            }
          }
        }
      })
      const wrapper = createWrapper({ store, })
      await flushPromises()

      wrapper.vm.atStart( event )

      expect( setTemporaryItem ).toBeCalled()
      expect( setTemporaryItem ).toBeCalledWith(
        expect.anything(),
        { tmpItem: store.state.items.baseItemsList[ event.oldIndex ] },
        void 0
      )
    })

    test('`atEnd` should set `data.draggedItem` to empty string', async () => {
      const event = {
        item: {
          attributes: [{ nodeValue: 'logo' }],
        },
        oldIndex: 0
      }
      const setTemporaryItem = jest.fn()
      const store = createStore({
        modules: {
          items: {
            actions: {
              setTemporaryItem
            }
          }
        }
      })
      const wrapper = createWrapper({ store, })
      await flushPromises()

      wrapper.vm.atStart( event )

      expect( wrapper.vm.draggedItem.length ).not.toBe( 0 )

      wrapper.vm.atEnd()

      expect( wrapper.vm.draggedItem ).toBe( '' )
    })

    test('`baseItemsList` should return items from store.state.items', async () => {
      const store   = createStore()
      const wrapper = createWrapper({ store, })
      await flushPromises()

      expect( wrapper.vm.baseItemsList ).toEqual( store.state.items.baseItemsList )
    })

    test('`baseItemsList` should accept a value and do nothing with the payload', async () => {
      const store   = createStore()
      const wrapper = createWrapper({ store, })
      await flushPromises()
      wrapper.vm.baseItemsList = [ 'some', 'thing' ]

      expect( wrapper.vm.baseItemsList ).toEqual( store.state.items.baseItemsList )
    })

  })

  test('`draggable` should render as much BaseItems as exists in baseItemsList', async () => {
    const store   = createStore()
    const wrapper = createWrapper({ store, })

    expect( wrapper.findAll('baseitem-stub').length ).toBe( store.state.items.baseItemsList.length )
  })

})