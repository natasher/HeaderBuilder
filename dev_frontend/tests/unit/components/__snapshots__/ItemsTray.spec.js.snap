// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ItemsTray: snapshots: should match if item is dragged 1`] = `
<section
  id="mfn-hb-items"
>
  <div
    class="header"
  >
    <span
      class="title"
    >
      
      Drag item and drop into workspace
    
    </span>
  </div>
   
  <draggable-stub
    class="hb-items clearfix"
    clone="function _default(original) {
          return original;
        }"
    element="div"
    options="[object Object]"
    value="[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object]"
  >
    <baseitem-stub
      item="[object Object]"
    />
    <baseitem-stub
      item="[object Object]"
    />
    <baseitem-stub
      item="[object Object]"
    />
    <baseitem-stub
      item="[object Object]"
      placeholder="true"
    />
    <baseitem-stub
      item="[object Object]"
    />
    <baseitem-stub
      item="[object Object]"
    />
    <baseitem-stub
      item="[object Object]"
    />
    <baseitem-stub
      item="[object Object]"
    />
    <baseitem-stub
      item="[object Object]"
    />
  </draggable-stub>
</section>
`;

exports[`ItemsTray: snapshots: should match when no item is dragged 1`] = `
<section
  id="mfn-hb-items"
>
  <div
    class="header"
  >
    <span
      class="title"
    >
      
      Drag item and drop into workspace
    
    </span>
  </div>
   
  <draggable-stub
    class="hb-items clearfix"
    clone="function _default(original) {
          return original;
        }"
    element="div"
    options="[object Object]"
    value="[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object]"
  >
    <baseitem-stub
      item="[object Object]"
    />
    <baseitem-stub
      item="[object Object]"
    />
    <baseitem-stub
      item="[object Object]"
    />
    <baseitem-stub
      item="[object Object]"
    />
    <baseitem-stub
      item="[object Object]"
    />
    <baseitem-stub
      item="[object Object]"
    />
    <baseitem-stub
      item="[object Object]"
    />
    <baseitem-stub
      item="[object Object]"
    />
    <baseitem-stub
      item="[object Object]"
    />
  </draggable-stub>
</section>
`;
